Ensure "watchers" group is set to watch any jiras created

eriknelson opened the pull request:
Fixes #13

eriknelson committed:
Update to use ES2020

Allows for the usage of Promise.allSettled

eriknelson committed:
Add ensureWatchers feature to updateIssue

* Encapsulates watcher logic in its own class

eriknelson committed:
Add compiled js

eriknelson committed:
Add watcher ensure to createIssue

djzager reviewed (changes requested):
Only thing I see that needs to happen is to make the desiredWatchers an optional input to the reconcile issue action.

eriknelson commented on src/watchers.ts:
```diff
@@ -0,0 +1,130 @@
+    // and often there's very little information (if any) about why
+    // 404 - Just a bad URL
+    // 415 - Missing Content-Type header
+    const results = await Promise.allSettled(
```
+1 that's a good idea, I'll add. Thank you.

djzager replied:
Yeah I was suggesting a softer warn. It really depends how often we expect it to fail trying to do this. Don't want all of the workflows failing because it failed to add a watcher and succeeded doing everything else.

eriknelson replied:
@djzager I think you're right about that, going to update.

eriknelson commented on src/watchers.ts:
```diff
@@ -0,0 +1,130 @@
+    // other than themselves. Need to dig into this if we want to add other users
+    // here. The exact permissions necessary should be documented.
+    this.desiredWatchers = [
+      'ernelson@redhat.com',
```
+1, much more flexible. Let's catch up on how to do that. still learning actions myself a bit.

eriknelson commented on src/watchers.ts:
```diff
@@ -0,0 +1,130 @@
+
+    // Annoyingly Typescript has no way to understand what's in the results
+    // array, so this actually requires a cast
+    const failures = (results.filter(res =>res.status === 'rejected'
```
Let's agree on some type style enforcement tool and employ that with a separate PR so it's uniform.

eriknelson committed:
Parameterize the auto watch list from action input

djzager commented on reconcile-issue/action.yml:
```diff
@@ -26,6 +26,10 @@ inputs:
   additionalLabels:
     description: |
       Comma separated list of labels that should be added to Jira issues.
+  autoWatchAccounts:
```
nit but `additionalWatchers` might convey the same thing with less confusion.

eriknelson replied:
"additional" is a little weird to me, because what is it in addition to?

autoWatchAccounts are the accounts that should be automatically watching the synced Jira.

djzager replied:
In addition to the bot that is automatically added as watcher. Prefer no confusion :sunglasses: 
If that's true, why not just `watchers`?

eriknelson replied:
I'm expecting to add a field that's the opposite polarity. So autoWatchAccounts bring the declarative "these are the accounts that must be watching" and then I plan to add "autoUnwatchAccounts".

In light of that do you have any good field names for both?

djzager commented on src/jira.ts:
```diff
@@ -303,6 +313,14 @@ class Jira {
     }
 
+    // Ensure watchers from watch group are subscribed for updated
+    const jwm = new JiraWatcherManager(jiraUrl, this.token);
```
I wonder, do you want to add these only to new issues and leave existing issues alone?

I suspect it works either way without any significant harm but wanted to mention it.

eriknelson replied:
I think I'm declaring the accounts that should be watching jiras spawned as a consequence of this bot by adding them to the list in the action. It should always be reconciling that state, even on existing issues IMO.

eriknelson committed:
Drop down a failed watcher addition to a warning

Probably shouldn't consider the entire action failed if a watcher
simply failed to get added.

djzager commented on src/watchers.ts:
```diff
@@ -0,0 +1,120 @@
+  private autoWatchAccounts: string[];
+
+  constructor(jiraIssueUrl: string, botToken: string) {
+    const autoWatchAccounts = core.getInput('autoWatchAccounts');
```
I don't have strong reasoning for this, but I liked having hall of the "inputs" to the reconcile-issue living at the very top of `reconcile-issue.ts` so it doesn't get lost.

Would it be possible to:
1. Accept the input in `reconcile-issue::reconcileIssue()`
2. Pass the watchers to Jira constructor
3. On create (and update) issue, the watchers are passed in when creating new JiraWatcherManagers.

eriknelson replied:
Yeah this is dependency injection and as we learned with the broker it can get really ugly when you plumb a dependency through as constructor arguments, especially as they get progressively more nested and pervasive (the "logger") issue.

I think your suggestion is a good one here though and I will update.

djzager replied:
after chatting about this one, I think it's a nack on this one (rightfully so). Marking as resolved.

djzager reviewed (commented):
None of these comments are "must have" but wanted to get your thoughts.

eriknelson committed:
Update watcher field name to addWatchers

Based on PR feedback, agreed upon name
Plan to do dropWatchers in a follow up PR

eriknelson committed:
Adjust some of the logs for noise
